pkg/controllers/cassandra/actions/scalein_test.go

```diff
@@ -0,0 +1,231 @@
+		expectedErr         bool
+		mutator             func(*framework.StateFixture)
+	}
+	tests := map[string]testT{
+		"Error if StatefulSet not listed": {
+			cluster: generate.CassandraClusterConfig{
+				Name:      "cluster1",
+				Namespace: "ns1",
+			},
+			nodePool: generate.CassandraClusterNodePoolConfig{
+				Name:     "pool1",
+				Replicas: 123,
+			},
+			expectedErr: true,
+		},
+		"Error if clientset.Update fails (e.g. listed but not found)": {
+			kubeObjects: []runtime.Object{
+				generate.StatefulSet(
+					generate.StatefulSetConfig{
+						Name:      "cass-cluster1-pool1",
+						Namespace: "ns1",
+						Replicas:  ptr.Int32(122),
+					},
+				),
+			},
+			cluster: generate.CassandraClusterConfig{
+				Name:      "cluster1",
+				Namespace: "ns1",
+			},
+			nodePool: generate.CassandraClusterNodePoolConfig{
+				Name:     "pool1",
+				Replicas: 123,
+			},
+			expectedErr: true,
+			mutator: func(f *framework.StateFixture) {
+				err := f.KubeClient().
+					AppsV1beta1().
+					StatefulSets("ns1").
+					Delete("cass-cluster1-pool1", &metav1.DeleteOptions{})
+				if err != nil {
+					f.T.Fatal(err)
+				}
+			},
+		},
+		"Idempotent: No error if ReplicaCount already matches the actual ReplicaCount": {
+			kubeObjects: []runtime.Object{
+				generate.StatefulSet(
+					generate.StatefulSetConfig{
+						Name:      "cass-cluster1-pool1",
+						Namespace: "ns1",
+						Replicas:  ptr.Int32(124),
+					},
+				),
+			},
+			cluster: generate.CassandraClusterConfig{
+				Name:      "cluster1",
+				Namespace: "ns1",
+			},
+			nodePool: generate.CassandraClusterNodePoolConfig{
+				Name:     "pool1",
+				Replicas: 124,
+			},
+			expectedStatefulSet: &generate.StatefulSetConfig{
+				Name:      "cass-cluster1-pool1",
+				Namespace: "ns1",
+				Replicas:  ptr.Int32(124),
+			},
+			expectedErr: false,
+		},
+		"The replicas count is decremented without successfully decommissioned pilots": {
+			kubeObjects: []runtime.Object{
+				generate.StatefulSet(
+					generate.StatefulSetConfig{
+						Name:      "cass-cluster1-pool1",
+						Namespace: "ns1",
+						Replicas:  ptr.Int32(2),
+					},
+				),
+			},
+			navObjects: []runtime.Object{
+				generate.CassPilot(
+					generate.PilotConfig{
+						Name:      "cass-cluster1-pool1-0",
+						Namespace: "ns1",
+						NodePool:  "pool1",
+						Cluster:   "cluster1",
+					},
+				),
+				generate.CassPilot(
+					generate.PilotConfig{
+						Name:      "cass-cluster1-pool1-1",
+						Namespace: "ns1",
+						NodePool:  "pool1",
+						Cluster:   "cluster1",
+					},
+				),
+			},
+			cluster: generate.CassandraClusterConfig{
+				Name:      "cluster1",
+				Namespace: "ns1",
+			},
+			nodePool: generate.CassandraClusterNodePoolConfig{
+				Name:     "pool1",
+				Replicas: 1,
+			},
+			expectedStatefulSet: &generate.StatefulSetConfig{
+				Name:      "cass-cluster1-pool1",
+				Namespace: "ns1",
+				Replicas:  ptr.Int32(2),
+			},
```

Comment:
Perhaps there should also be an assertion here to demonstrate that one of the pilots gets marked as decommissioned.